packages/repository-tests/src/crud/relations/integration/has-many.factory.integration.ts

```diff
@@ -0,0 +1,259 @@
+      ).find();
+
+      const persistedReviewsApprovedByCustomerOne = await reviewRepo.find({
+        where: {
+          approvedId: existingCustomerId,
+        },
+      });
+      const persistedReviewsApprovedByCustomerTwo = await reviewRepo.find({
+        where: {
+          approvedId: existingCustomerId + 1,
+        },
+      });
+
+      expect(reviewsApprovedByCustomerOne).to.eql(
+        persistedReviewsApprovedByCustomerOne,
+      );
+      expect(reviewsApprovedByCustomerTwo).to.eql(
+        persistedReviewsApprovedByCustomerTwo,
+      );
+    });
+
+    //--- HELPERS ---//
+
+    async function givenPersistedCustomerInstance() {
+      const customer = await customerRepo.create({name: 'a customer'});
+      existingCustomerId = customer.id;
+    }
+
+    function givenConstrainedRepositories() {
+      const orderFactoryFn = createHasManyRepositoryFactory<
+        Order,
+        typeof Order.prototype.id,
+        typeof Customer.prototype.id
+      >(
+        Customer.definition.relations.orders as HasManyDefinition,
+        Getter.fromValue(orderRepo),
+      );
+
+      customerOrderRepo = orderFactoryFn(existingCustomerId);
+    }
+
+    function givenRepositoryFactoryFunctions() {
+      customerAuthoredReviewFactoryFn = createHasManyRepositoryFactory(
+        Customer.definition.relations.reviewsAuthored as HasManyDefinition,
+        Getter.fromValue(reviewRepo),
+      );
+      customerApprovedReviewFactoryFn = createHasManyRepositoryFactory(
+        Customer.definition.relations.reviewsApproved as HasManyDefinition,
+        Getter.fromValue(reviewRepo),
+      );
+    }
+  });
+
+  function givenCrudRepositories() {
+    customerRepo = new repositoryClass(Customer, db);
+    orderRepo = new repositoryClass(Order, db);
+    reviewRepo = new repositoryClass(Review, db);
+  }
+
+  //--- HELPERS ---//
+
+  class Order extends Entity {
+    id: string;
+    description: string;
+    customerId: string;
+
+    static definition = new ModelDefinition('Order')
+      .addProperty('id', {type: features.idType, id: true, generated: true})
+      .addProperty('description', {type: 'string', required: true})
+      .addProperty('customerId', {type: 'string'})
+      .addRelation({
+        name: 'customer',
+        type: RelationType.belongsTo,
+        source: Order,
+        target: () => Customer,
+        keyFrom: 'customerId',
+        keyTo: 'id',
+      });
+  }
+
+  class Review extends Entity {
+    id: string;
+    description: string;
+    authorId: string;
+    approvedId: string;
+
+    static definition = new ModelDefinition('Review')
+      .addProperty('id', {type: features.idType, id: true, generated: true})
+      .addProperty('description', {type: 'string', required: true})
+      .addProperty('authorId', {type: features.idType, required: false})
+      .addProperty('approvedId', {type: features.idType, required: false});
+  }
+
+  class Customer extends Entity {
+    id: string;
```

Comment:
Ditto, let's use `mixedIdType` or `unknown`. Same comment applies to all other similar model-property definitions.